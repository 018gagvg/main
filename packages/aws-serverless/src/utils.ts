import type { Scope } from '@sentry/types';
import { addExceptionMechanism } from '@sentry/utils';

/**
 * Marks an event as unhandled by adding a span processor to the passed scope.
 */
export function markEventUnhandled(scope: Scope): Scope {
  scope.addEventProcessor(event => {
    addExceptionMechanism(event, { handled: false });
    return event;
  });

  return scope;
}
