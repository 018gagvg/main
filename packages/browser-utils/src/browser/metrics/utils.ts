import type { SentrySpan } from '@sentry/core';
import { spanToJSON, startInactiveSpan, withActiveSpan } from '@sentry/core';
import type { Span, SpanTimeInput, StartSpanOptions } from '@sentry/types';

/**
 * Checks if a given value is a valid measurement value.
 */
export function isMeasurementValue(value: unknown): value is number {
  return typeof value === 'number' && isFinite(value);
}

/**
 * Helper function to start child on transactions. This function will make sure that the transaction will
 * use the start timestamp of the created child span if it is earlier than the transactions actual
 * start timestamp.
 *
 * Note: this will not be possible anymore in v8,
 * unless we do some special handling for browser here...
 */
export function startAndEndSpan(
  parentSpan: Span,
  startTimeInSeconds: number,
  endTime: SpanTimeInput,
  { ...ctx }: StartSpanOptions,
): Span | undefined {
  const parentStartTime = spanToJSON(parentSpan).start_timestamp;
  if (parentStartTime && parentStartTime > startTimeInSeconds) {
    // We can only do this for SentrySpans...
    if (typeof (parentSpan as Partial<SentrySpan>).updateStartTime === 'function') {
      (parentSpan as SentrySpan).updateStartTime(startTimeInSeconds);
    }
  }

  // The return value only exists for tests
  return withActiveSpan(parentSpan, () => {
    const span = startInactiveSpan({
      startTime: startTimeInSeconds,
      ...ctx,
    });

    if (span) {
      span.end(endTime);
    }

    return span;
  });
}
