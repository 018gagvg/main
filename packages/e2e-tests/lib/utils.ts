/* eslint-disable no-console */
import * as childProcess from 'child_process';

// https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message
export function printCIErrorMessage(message: string): void {
  if (process.env.CI) {
    console.log(`::error::${message}`);
  } else {
    console.log(message);
  }
}

interface SpawnAsync {
  stdout: string;
  stderr: string;
  error?: Error;
  status: number | null;
}

interface SpawnOptions {
  timeout?: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  env?: any;
  cwd?: string;
}

export function spawnAsync(cmd: string, options?: SpawnOptions, input?: string): Promise<SpawnAsync> {
  const timeoutMs = options?.timeout || 60_000 * 5;

  return new Promise<SpawnAsync>(resolve => {
    const cp = childProcess.spawn(cmd, { shell: true, ...options });

    // Ensure we properly time out after max. 5 min per command
    let timeout: ReturnType<typeof setTimeout> | undefined = setTimeout(() => {
      console.log(`Command "${cmd}" timed out after 5 minutes.`);
      cp.kill();
      end(null, `ETDIMEDOUT: Process timed out after ${timeoutMs} ms.`);
    }, timeoutMs);

    const stderr: unknown[] = [];
    const stdout: string[] = [];
    let error: Error | undefined;

    function end(status: number | null, errorMessage?: string): void {
      // This means we already ended
      if (!timeout) {
        return;
      }

      if (!error && errorMessage) {
        error = new Error(errorMessage);
      }

      clearTimeout(timeout);
      timeout = undefined;
      resolve({
        stdout: stdout.join(''),
        stderr: stderr.join(''),
        error: error || (status !== 0 ? new Error(`Process exited with status ${status}`) : undefined),
        status,
      });
    }

    cp.stdout.on('data', data => {
      stdout.push(data ? (data as object).toString() : '');
    });

    cp.stderr.on('data', data => {
      stderr.push(data ? (data as object).toString() : '');
    });

    cp.on('error', e => {
      error = e;
    });

    cp.on('close', status => {
      end(status);
    });

    if (input) {
      cp.stdin.write(input);
      cp.stdin.end();
    }
  });
}

export function prefixObjectKeys(
  obj: Record<string, string | undefined>,
  prefix: string,
): Record<string, string | undefined> {
  return Object.keys(obj).reduce<Record<string, string | undefined>>((result, key) => {
    result[prefix + key] = obj[key];
    return result;
  }, {});
}
