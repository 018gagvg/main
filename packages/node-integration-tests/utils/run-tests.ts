/* eslint-disable no-console */
import childProcess from 'child_process';
import os from 'os';

const testPaths = childProcess.execSync('jest --listTests', { encoding: 'utf8' }).trim().split('\n');

let testsSucceeded = true;
const testAmount = testPaths.length;
const fails: string[] = [];

const threads = os.cpus().map(async (_, i) => {
  while (testPaths.length > 0) {
    const testPath = testPaths.pop();
    console.log(`(Worker ${i}) Running test "${testPath}"`);
    await new Promise(resolve => {
      const jestProcess = childProcess.spawn('jest', ['--runTestsByPath', testPath as string, '--forceExit']);

      let output = '';

      jestProcess.stdout.on('data', (data: Buffer) => {
        output = output + data.toString();
      });

      jestProcess.stderr.on('data', (data: Buffer) => {
        output = output + data.toString();
      });

      jestProcess.on('error', error => {
        console.log(`(Worker ${i}) Error in test "${testPath}"`, error);
        console.log(output);
        resolve();
      });

      jestProcess.on('exit', exitcode => {
        console.log(`(Worker ${i}) Finished test "${testPath}"`);
        console.log(output);
        if (exitcode !== 0) {
          fails.push(`FAILED: "${testPath}"`);
          testsSucceeded = false;
        }
        resolve();
      });
    });
  }
});

void Promise.all(threads).then(() => {
  console.log('-------------------');
  console.log(`Successfully ran ${testAmount} tests.`);
  if (!testsSucceeded) {
    console.log('Not all tests succeeded:\n');
    fails.forEach(fail => {
      console.log(`‚óè ${fail}`);
    });
    process.exit(1);
  } else {
    console.log('All tests succeeded.');
    process.exit(0);
  }
});
