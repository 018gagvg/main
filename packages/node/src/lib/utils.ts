import { mkdir, mkdirSync, statSync } from 'fs';
import { dirname, join, resolve } from 'path';

const _0777 = parseInt('0777', 8);

function mkdirAsync(path: string, mode: number): Promise<void> {
  // We cannot use util.promisify here because that was only introduced
  // in Node 8 and we need to support older Node versions.
  return new Promise((res, reject) => {
    mkdir(path, mode, err => (err ? reject(err) : res()));
  });
}

/**
 * Recursively creates the given path.
 *
 * @param path A relative or absolute path to create.
 * @returns A promise that resolves when the path has been created.
 */
export async function mkdirp(path: string): Promise<void> {
  // tslint:disable-next-line:no-bitwise
  const mode = _0777 & ~process.umask();
  const realPath = resolve(path);

  try {
    return mkdirAsync(realPath, mode);
  } catch (err) {
    if (err.code === 'ENOENT') {
      await mkdirp(dirname(realPath));
      return mkdirAsync(realPath, mode);
    } else {
      try {
        if (!statSync(realPath).isDirectory()) {
          throw err;
        }
      } catch (_) {
        throw err;
      }
    }
  }
}

/**
 * Synchronous version of {@link mkdirp}.
 *
 * @param path A relative or absolute path to create.
 */
export function mkdirpSync(path: string): void {
  // tslint:disable-next-line:no-bitwise
  const mode = _0777 & ~process.umask();
  const realPath = resolve(path);

  try {
    mkdirSync(realPath, mode);
  } catch (err) {
    if (err.code === 'ENOENT') {
      mkdirpSync(dirname(realPath));
      mkdirSync(realPath, mode);
    } else {
      try {
        if (!statSync(realPath).isDirectory()) {
          throw err;
        }
      } catch (_) {
        throw err;
      }
    }
  }
}
