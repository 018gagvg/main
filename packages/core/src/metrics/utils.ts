import type { MeasurementUnit, Primitive } from '@sentry/types';
import { dropUndefinedKeys } from '@sentry/utils';
import type { MetricType } from './types';

/**
 * Generate bucket key from metric properties.
 */
export function getBucketKey(
  metricType: MetricType,
  name: string,
  unit: MeasurementUnit,
  tags: { [key: string]: string },
): string {
  const stringifiedTags = Object.entries(dropUndefinedKeys(tags)).sort() as unknown as string[];
  return [metricType, name, unit].concat(stringifiedTags).join('');
}
