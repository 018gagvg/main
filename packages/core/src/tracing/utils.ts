import type { Span, Transaction } from '@sentry/types';
import type { Scope } from '@sentry/types';
import { addNonEnumerableProperty } from '@sentry/utils';
import { getCurrentScope } from '../currentScopes';

import type { Hub } from '../hub';
import { getCurrentHub } from '../hub';

/**
 * Grabs active transaction off scope.
 *
 * @deprecated You should not rely on the transaction, but just use `startSpan()` APIs instead.
 */
export function getActiveTransaction<T extends Transaction>(maybeHub?: Hub): T | undefined {
  // eslint-disable-next-line deprecation/deprecation
  const hub = maybeHub || getCurrentHub();
  // eslint-disable-next-line deprecation/deprecation
  const scope = hub.getScope();
  // eslint-disable-next-line deprecation/deprecation
  return scope.getTransaction() as T | undefined;
}

// so it can be used in manual instrumentation without necessitating a hard dependency on @sentry/utils
export { stripUrlQueryAndFragment } from '@sentry/utils';

/**
 * Returns the currently active span.
 */
export function getActiveSpan(): Span | undefined {
  // eslint-disable-next-line deprecation/deprecation
  return getCurrentScope().getSpan();
}

const SCOPE_ON_START_SPAN_FIELD = '_sentryScope';
const ISOLATION_SCOPE_ON_START_SPAN_FIELD = '_sentryIsolationScope';

type SpanWithScopes = Span & {
  [SCOPE_ON_START_SPAN_FIELD]?: Scope;
  [ISOLATION_SCOPE_ON_START_SPAN_FIELD]?: Scope;
};

/** Store the scope & isolation scope for a span, which can the be used when it is finished. */
export function setCapturedScopesOnSpan(span: Span | undefined, scope: Scope, isolationScope: Scope): void {
  if (span) {
    addNonEnumerableProperty(span, ISOLATION_SCOPE_ON_START_SPAN_FIELD, isolationScope);
    addNonEnumerableProperty(span, SCOPE_ON_START_SPAN_FIELD, scope);
  }
}

/**
 * Grabs the scope and isolation scope off a span that were active when the span was started.
 */
export function getCapturedScopesOnSpan(span: Span): { scope?: Scope; isolationScope?: Scope } {
  return {
    scope: (span as SpanWithScopes)[SCOPE_ON_START_SPAN_FIELD],
    isolationScope: (span as SpanWithScopes)[ISOLATION_SCOPE_ON_START_SPAN_FIELD],
  };
}
