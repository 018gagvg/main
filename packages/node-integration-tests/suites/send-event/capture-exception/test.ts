import { getEventRequest, runServer, updateForSnapshot } from '../../../utils';

test('should send captureException', async () => {
  const url = await runServer(__dirname);
  const requestBody = await getEventRequest(url);

  expect(updateForSnapshot(requestBody)).toMatchSnapshot();
});
