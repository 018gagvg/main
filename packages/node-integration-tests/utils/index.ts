/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { logger, parseSemver } from '@sentry/utils';
import axios from 'axios';
import { Express } from 'express';
import * as http from 'http';
import { RequestOptions } from 'https';
import nock from 'nock';
import * as path from 'path';
import { getPortPromise } from 'portfinder';

export type TestServerConfig = {
  url: string;
  server: http.Server;
  scope: nock.Scope;
};

/**
 * Returns`describe` or `describe.skip` depending on allowed major versions of Node.
 *
 * @param {{ min?: number; max?: number }} allowedVersion
 * @return {*}  {jest.Describe}
 */
export const conditionalTest = (allowedVersion: { min?: number; max?: number }): jest.Describe => {
  const NODE_VERSION = parseSemver(process.versions.node).major;
  if (!NODE_VERSION) {
    return describe.skip;
  }

  return NODE_VERSION < (allowedVersion.min || -Infinity) || NODE_VERSION > (allowedVersion.max || Infinity)
    ? describe.skip
    : describe;
};

/**
 * Asserts against a Sentry Event ignoring non-deterministic properties
 *
 * @param {Record<string, unknown>} actual
 * @param {Record<string, unknown>} expected
 */
export const assertSentryEvent = (actual: Record<string, unknown>, expected: Record<string, unknown>): void => {
  expect(actual).toMatchObject({
    event_id: expect.any(String),
    timestamp: expect.anything(),
    ...expected,
  });
};

/**
 * Asserts against a Sentry Transaction ignoring non-deterministic properties
 *
 * @param {Record<string, unknown>} actual
 * @param {Record<string, unknown>} expected
 */
export const assertSentryTransaction = (actual: Record<string, unknown>, expected: Record<string, unknown>): void => {
  expect(actual).toMatchObject({
    event_id: expect.any(String),
    timestamp: expect.anything(),
    start_timestamp: expect.anything(),
    spans: expect.any(Array),
    type: 'transaction',
    ...expected,
  });
};

/**
 * Parses response body containing an Envelope
 *
 * @param {string} body
 * @return {*}  {Array<Record<string, unknown>>}
 */
export const parseEnvelope = (body: string): Array<Record<string, unknown>> => {
  return body.split('\n').map(e => JSON.parse(e));
};

/**
 * Intercepts and extracts up to a number of requests containing Sentry envelopes.
 *
 * @param config The url, server instance and the nock scope.
 * @param count The expected amount of requests to the envelope endpoint. If
 * the amount of sentrequests is lower than`count`, this function will not resolve.
 * @param method The method of the request. Defaults to `GET`.
 * @param endServer: Whether to stop the server after the requests have been intercepted.
 * @returns The intercepted envelopes.
 */
export const getMultipleEnvelopeRequest = async (
  config: TestServerConfig,
  count: number,
  method: 'get' | 'post' = 'get',
  endServer: boolean = true,
): Promise<Record<string, unknown>[][]> => {
  // eslint-disable-next-line no-async-promise-executor
  return (
    await Promise.all([setupNock(config.scope, config.server, count, endServer), makeRequest(method, config.url)])
  )[0];
};

const setupNock = async (
  scope: nock.Scope,
  server: http.Server,
  count: number,
  endServer: boolean,
): Promise<Record<string, unknown>[][]> => {
  return new Promise(resolve => {
    const envelopes: Record<string, unknown>[][] = [];

    scope
      .post('/api/1337/envelope/', body => {
        const envelope = parseEnvelope(body);

        envelopes.push(envelope);

        if (count === envelopes.length) {
          if (endServer) {
            scope.persist(false);
            server.close(() => {
              resolve(envelopes);
            });
          }
          resolve(envelopes);
        }

        return true;
      })
      .query(true) // accept any query params - used for sentry_key param
      .reply(200);
  });
};

const makeRequest = async (method: 'get' | 'post', url: string): Promise<void> => {
  try {
    if (method === 'get') {
      await axios.get(url);
    } else {
      await axios.post(url);
    }
  } catch (e) {
    // We sometimes expect the request to fail, but not the test.
    // So, we do nothing.
    logger.warn(e);
  }
};

/**
 * Sends a get request to given URL, with optional headers
 *
 * @param {URL} url
 * @param {Record<string, string>} [headers]
 * @return {*}  {Promise<any>}
 */
export const getAPIResponse = async (config: TestServerConfig, headers?: Record<string, string>): Promise<unknown> => {
  return new Promise(resolve => {
    const url = new URL(config.url);

    http.get(
      headers
        ? ({
            protocol: url.protocol,
            host: url.hostname,
            path: url.pathname,
            port: url.port,
            headers,
          } as RequestOptions)
        : url,
      response => {
        let body = '';

        response.on('data', function (chunk: string) {
          body += chunk;
        });
        response.on('end', function () {
          config.scope.persist(false);
          config.server.close(() => {
            resolve(JSON.parse(body));
          });
        });
      },
    );
  });
};

/**
 * Intercepts and extracts a single request containing a Sentry envelope
 *
 * @param url The url the intercepted request will be directed to.
 * @param method The method of the request. Defaults to `GET`.
 * @returns The extracted envelope.
 */
export const getEnvelopeRequest = async (
  config: TestServerConfig,
  method: 'get' | 'post' = 'get',
  endServer: boolean = true,
): Promise<Array<Record<string, unknown>>> => {
  return (await getMultipleEnvelopeRequest(config, 1, method, endServer))[0];
};

/**
 * Runs a test server
 *
 * @param {string} testDir
 * @param {string} [serverPath]
 * @param {string} [scenarioPath]
 * @return {*}  {Promise<string>}
 */
export async function runServer(
  testDir: string,
  serverPath?: string,
  scenarioPath?: string,
): Promise<TestServerConfig> {
  const port = await getPortPromise();
  const url = `http://localhost:${port}/test`;
  const defaultServerPath = path.resolve(process.cwd(), 'utils', 'defaults', 'server');

  const { server, scope } = await new Promise<{ server: http.Server; scope: nock.Scope }>(resolve => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-member-access
    const app = require(serverPath || defaultServerPath).default as Express;

    app.get('/test', (_req, res) => {
      try {
        require(scenarioPath || `${testDir}/scenario`);
      } finally {
        res.status(200).end();
      }
    });

    const server = app.listen(port, () => {
      const scope = nock('https://dsn.ingest.sentry.io').persist();

      resolve({ server, scope });
    });
  });

  return { url, server, scope };
}

export async function runScenario(serverUrl: string): Promise<void> {
  return new Promise<void>(resolve => {
    http
      .get(serverUrl, res => {
        res.on('data', () => undefined);
        res.on('end', resolve);
      })
      .end();
  });
}
