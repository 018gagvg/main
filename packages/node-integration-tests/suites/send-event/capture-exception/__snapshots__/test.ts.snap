// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should send captureException 1`] = `
Object {
  "environment": "production",
  "event_id": Any<String>,
  "exception": Object {
    "values": Array [
      Object {
        "mechanism": Object {
          "handled": true,
          "synthetic": true,
          "type": "generic",
        },
        "stacktrace": Object {
          "frames": Array [
            Object {
              "colno": 75,
              "context_line": "            step((generator = generator.apply(thisArg, _arguments || [])).next());
",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/tslib/tslib.js",
              "function": "undefined.<anonymous>",
              "in_app": false,
              "lineno": 115,
              "module": "tslib:tslib",
              "post_context": Array [
                "        });
",
                "    };
",
                "
",
                "    __generator = function (thisArg, body) {
",
                "        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
",
                "        return g = { next: verb(0), \\"throw\\": verb(1), \\"return\\": verb(2) }, typeof Symbol === \\"function\\" && (g[Symbol.iterator] = function()  {snip}",
                "        function verb(n) { return function (v) { return step([n, v]); }; }
",
              ],
              "pre_context": Array [
                "
",
                "    __awaiter = function (thisArg, _arguments, P, generator) {
",
                "        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
",
                "        return new (P || (P = Promise))(function (resolve, reject) {
",
                "            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
",
                "            function rejected(value) { try { step(generator[\\"throw\\"](value)); } catch (e) { reject(e); } }
",
                "            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
",
              ],
            },
            Object {
              "colno": 57,
              "context_line": "        function verb(n) { return function (v) { return step([n, v]); }; }
",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/tslib/tslib.js",
              "function": "Object.next",
              "in_app": false,
              "lineno": 122,
              "module": "tslib:tslib",
              "post_context": Array [
                "        function step(op) {
",
                "            if (f) throw new TypeError(\\"Generator is already executing.\\");
",
                "            while (_) try {
",
                "                if (f = 1, y && (t = op[0] & 2 ? y[\\"return\\"] : op[0] ? y[\\"throw\\"] || ((t = y[\\"return\\"]) && t.call(y), 0) : y.next) && !(t =  {snip}",
                "                if (y = 0, t) op = [op[0] & 2, t.value];
",
                "                switch (op[0]) {
",
                "                    case 0: case 1: t = op; break;
",
              ],
              "pre_context": Array [
                "            step((generator = generator.apply(thisArg, _arguments || [])).next());
",
                "        });
",
                "    };
",
                "
",
                "    __generator = function (thisArg, body) {
",
                "        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
",
                "        return g = { next: verb(0), \\"throw\\": verb(1), \\"return\\": verb(2) }, typeof Symbol === \\"function\\" && (g[Symbol.iterator] = function()  {snip}",
              ],
            },
            Object {
              "colno": 27,
              "context_line": "                op = body.call(thisArg, _);
",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/tslib/tslib.js",
              "function": "step",
              "in_app": false,
              "lineno": 141,
              "module": "tslib:tslib",
              "post_context": Array [
                "            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
",
                "            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
",
                "        }
",
                "    };
",
                "
",
                "    __createBinding = function(o, m, k, k2) {
",
                "        if (k2 === undefined) k2 = k;
",
              ],
              "pre_context": Array [
                "                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
",
                "                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
",
                "                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
",
                "                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
",
                "                        if (t[2]) _.ops.pop();
",
                "                        _.trys.pop(); continue;
",
                "                }
",
              ],
            },
            Object {
              "colno": 7,
              "context_line": "      require(\`\${testDir}/subject\`);",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/packages/node-integration-tests/utils/index.ts",
              "function": "Object.<anonymous>",
              "in_app": true,
              "lineno": 79,
              "module": "index.ts",
              "post_context": Array [
                "",
                "      setTimeout(() => server.close(), 500);",
                "    });",
                "",
                "    const server = app.listen(port, () => {",
                "      resolve();",
                "    });",
              ],
              "pre_context": Array [
                "",
                "  await new Promise(resolve => {",
                "    // eslint-disable-next-line @typescript-eslint/no-var-requires",
                "    const app = require(\`\${testDir}/server\`).default as Express;",
                "",
                "    app.get('/test', async () => {",
                "      require(\`\${testDir}/init\`);",
              ],
            },
            Object {
              "colno": 21,
              "context_line": "        return this.requireModule(from, moduleName);",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/jest-runtime/build/index.js",
              "function": "Runtime.requireModuleOrMock",
              "in_app": false,
              "lineno": 598,
              "module": "jest-runtime.build:index",
              "post_context": Array [
                "      }",
                "    } catch (e) {",
                "      if (e.code === 'MODULE_NOT_FOUND') {",
                "        const appendedMessage = (0, _helpers.findSiblingsWithFileExtension)(",
                "          this._config.moduleFileExtensions,",
                "          from,",
                "          moduleName",
              ],
              "pre_context": Array [
                "  }",
                "",
                "  requireModuleOrMock(from, moduleName) {",
                "    try {",
                "      if (this._shouldMock(from, moduleName)) {",
                "        return this.requireMock(from, moduleName);",
                "      } else {",
              ],
            },
            Object {
              "colno": 10,
              "context_line": "    this._loadModule(",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/jest-runtime/build/index.js",
              "function": "Runtime.requireModule",
              "in_app": false,
              "lineno": 433,
              "module": "jest-runtime.build:index",
              "post_context": Array [
                "      localModule,",
                "      from,",
                "      moduleName,",
                "      modulePath,",
                "      options,",
                "      moduleRegistry",
                "    );",
              ],
              "pre_context": Array [
                "      exports: {},",
                "      filename: modulePath,",
                "      id: modulePath,",
                "      loaded: false",
                "    };",
                "    moduleRegistry.set(modulePath, localModule);",
                "",
              ],
            },
            Object {
              "colno": 12,
              "context_line": "      this._execModule(localModule, options, moduleRegistry, fromPath);",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/jest-runtime/build/index.js",
              "function": "Runtime._loadModule",
              "in_app": false,
              "lineno": 577,
              "module": "jest-runtime.build:index",
              "post_context": Array [
                "    }",
                "",
                "    localModule.loaded = true;",
                "  }",
                "",
                "  _getFullTransformationOptions(options) {",
                "    return _objectSpread({}, options, {",
              ],
              "pre_context": Array [
                "      );",
                "    } else if (_path().default.extname(modulePath) === '.node') {",
                "      localModule.exports = require(modulePath);",
                "    } else {",
                "      // Only include the fromPath if a moduleName is given. Else treat as root.",
                "      const fromPath = moduleName ? from : null;",
                "",
              ],
            },
            Object {
              "colno": 68,
              "context_line": "    runScript[_transform().ScriptTransformer.EVAL_RESULT_VARIABLE].call(",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/node_modules/jest-runtime/build/index.js",
              "function": "Runtime._execModule",
              "in_app": false,
              "lineno": 867,
              "module": "jest-runtime.build:index",
              "post_context": Array [
                "      localModule.exports,",
                "      localModule, // module object",
                "      localModule.exports, // module exports",
                "      localModule.require, // require implementation",
                "      dirname, // __dirname",
                "      filename, // __filename",
                "      this._environment.global, // global object",
              ],
              "pre_context": Array [
                "        'You are trying to \`import\` a file after the Jest environment has been torn down.'",
                "      );",
                "",
                "      process.exitCode = 1;",
                "      return;",
                "    } //Wrapper",
                "",
              ],
            },
            Object {
              "colno": 8,
              "context_line": "Sentry.captureException('Error');",
              "filename": "/Users/onurtemizkan/Projects/sentry-javascript/packages/node-integration-tests/suites/send-event/capture-exception/subject.ts",
              "function": "Object.<anonymous>",
              "in_app": true,
              "lineno": 3,
              "module": "subject.ts",
              "post_context": Array [
                "",
              ],
              "pre_context": Array [
                "import * as Sentry from '@sentry/node';",
                "",
              ],
            },
          ],
        },
        "type": "Error",
        "value": "Error",
      },
    ],
  },
  "extra": Object {
    "normalizeDepth": 3,
  },
  "platform": "node",
  "sdk": Object {
    "integrations": Array [
      "InboundFilters",
      "FunctionToString",
      "ContextLines",
      "Console",
      "Http",
      "OnUncaughtException",
      "OnUnhandledRejection",
      "LinkedErrors",
    ],
    "name": "sentry.javascript.node",
    "packages": Array [
      Object {
        "name": "npm:@sentry/node",
        "version": "6.18.1",
      },
    ],
    "version": "6.18.1",
  },
  "tags": Object {
    "skippedNormalization": true,
  },
  "timestamp": Any<Number>,
}
`;
